Updating Workflow Execution concept page (#1015)
temporalio/documentation: docs/concepts/what-is-a-workflow-execution.md

id: what-is-a-workflow-execution
title: What is a Workflow Execution?
sidebar_label: Workflow Execution
description: A Temporal Workflow Execution is a durable, scalable, reliable, and reactive function execution. It is a Temporal Application’s main unit of execution.
tags:
  - explanation
---

A Temporal Workflow Execution is a durable, reliable, and scalable function execution.
It is a Temporal Application’s main unit of execution.

- [How to spawn a Workflow Execution in Go](/docs/go/how-to-spawn-a-workflow-execution-in-go)

> - [How to spawn a Workflow Execution in Java](/docs/java/how-to-spawn-a-workflow-execution-in-java)

Reviewer: Why only these two?

Author: Will add the equivalent links in the others.

> Each Temporal Workflow Execution has exclusive access to its local state, executes concurrently to all other Workflow Executions, and communicates with other Workflow Executions and the environment via message passing.

Reviewer: It communicates with the environment through activities only.

> A Temporal Application can consist of millions to billions of Workflow Executions.

Reviewer: A single workflow execution has limited size and throughput. An application scales out by executing more workflow executions in parallel.

Workflow Executions are lightweight components.

> A Workflow Execution consumes few compute resources; in fact, if a Workflow Execution is suspended, such as when it is in a waiting state, the Workflow Execution consumes no compute resources at all.

Reviewer: > A Workflow Execution consumes few compute resources;

I think this is confusing and not 100% correct. I can easily write a workflow execution that is going to consume all memory and a lot of CPU on a machine.

**Durability**

Durability is responsiveness over a period time.
A Workflow Execution is durable because it executes a Temporal Workflow Definition, also called a Temporal Workflow Function, your application code, effectively once and to completion—whether your code executes for seconds or years.

> Durability is realized through the execution semantics of a Workflow Execution, which are carried out via Temporal's Event loop.

Reviewer: This sounds confusing to me. 

I think durability is about persisting workflow execution state in the presence of various software and hardware failures.

Author: Per @dtornow - sounds like this is track Product is taking:
Durability is the absence of an imposed time limit.
Reliability is responsiveness in the presence of failure.
Scalability is responsiveness in the presence of load.

**Reliability**

Reliability is responsiveness in the presence of failure.
A Workflow Execution is reliable, because it is fully recoverable after a failure.
To be recoverable is the ability of a process to continue execution after the execution was suspended by a failure.
Reliability is realized through the execution semantics of a Workflow Execution, which are carried out via Temporal's Event loop.

**Scalability**

Scalability is responsiveness in the presence of load.

> A Workflow Execution is scalable because the Temporal Platform is capable of supporting millions to billions of them concurrently.

Reviewer: We need to mention here that a single workflow execution is **not** scalable.

This is realized in part due to the design and nature of a [Temporal Cluster](/docs/concepts/what-is-a-temporal-cluster) and in part due to the design and nature of [Worker Processes](/docs/concepts/what-is-a-worker-process).

### Commands

A Workflow Execution does just two things:

1. Issue [Commands](/docs/concepts/what-is-a-command)
2. Wait on an Awaitable provided from the generation of Commands.

> 

Reviewer: Commands and awaitables are not concepts that are directly exposed to the user. Is it possible to rewrite this section that is more concrete? For example, that workflow is allowed to call only Temporal APIs for blocking operations.

![Command generation and waiting](/diagrams/workflow-execution-progession-simple.svg)

Commands are issued by the use of Workflow APIs in the [Workflow Definition](/docs/concepts/what-is-a-workflow-definition).

Whenever the Workflow Function is executed, Commands are generated.
The Worker Process supervises the Command generation and makes sure that it corresponds to the current Event History (See [Deterministic constraints](/docs/concepts/what-is-a-workflow-definition/#deterministic-constraints)).
The Worker Process batches the Commands and then suspends progress to send the Commands to the Cluster whenever the Workflow Function reaches a place where it can no longer progress without a result from a Awaitable.

Awaitables are provided when using APIs for the following:

- Spawning a [Child Workflow Execution](/docs/concepts/what-is-a-child-workflow-execution)
- Sending a [Signal](/docs/concepts/what-is-a-signal)
- Spawning an [Activity Execution](/docs/concepts/what-is-an-activity-execution)
- Requesting cancellation of another Workflow Execution
- Starting a Timer

> 

Reviewer: What about workflow.Await or Future.get? These are not directly linked to any command.

- Terminated: The Workflow Execution was terminated.
- Cancelled: The Workflow Execution successfully handled a cancellation request.
- Failed: The Workflow Execution returned an error and failed.
- Timed Out: The Workflow Execution reached a timeout limit.

### Workflow Execution Chain

A Workflow Execution Chain is a sequence of Workflow Executions that share the same Workflow Id.
Each link in the Chain is often called a Workflow Run.
Each Workflow Run in the sequence is connected by one of the following:

- [Temporal Cron Job](/docs/concepts/what-is-a-temporal-cron-job)
- [Continue As New](/docs/concepts/what-is-continue-as-new)
* [Retries](/docs/concepts/what-is-a-retry-policy)

A Workflow Execution is uniquely identified by its [Namespace](/docs/concepts/what-is-a-namespace), [Workflow Id](/docs/concepts/what-is-a-workflow-id), and [Run Id](/docs/concepts/what-is-a-run-id).

The [Workflow Execution Timeout](/docs/concepts/what-is-a-workflow-execution-timeout) applies to a Workflow Execution Chain.
The [Workflow Run Timeout](/docs/concepts/what-is-a-workflow-run-timeout) applies to a single Workflow Execution (Workflow Run).

### Event loop

> A Workflow Execution is made up of a set of [Events](/docs/concepts/what-is-an-event) called an [Event History](/docs/concepts/what-is-an-event-history).

Reviewer 2: "...set of events..." - would calling this a '...sequence of events...' help contextualize it better as a log-based thing?

Events are created by the Temporal Cluster in response to either Commands or actions requested by a Temporal Client (such as a request to spawn a Workflow Execution).

![Workflow Execution](/diagrams/workflow-execution-swim-lane-01.svg)

### Time constraints

**Is there a limit to how long Workflows can run?**

No, there is no time constraint on how long a Workflow Execution can be Running.

However, Workflow Executions intended to run indefinitely should be written with some care.
The Temporal Cluster stores the complete Event History for the entire lifecycle of a Workflow Execution.

> There is a hard limit of 50,000 Events in a Workflow Execution Event History, as well as a hard limit of 50MB in terms of size.

Reviewer 2: Add here: "When the Event History reaches 50,000 Events or the size limit of 50MB, the Workflow Execution is forcefully terminated." because https://temporalio.slack.com/archives/CTRCR8RBP/p1647904244410309

> There is a hard limit of 50,000 Events in a Workflow Execution Event History, as well as a hard limit of 50MB in terms of size.

Reviewer 3: ```suggestion
There is a hard limit of 50,000 Events in a Workflow Execution Event History, as well as a hard limit of 50 MB in terms of size.
```

The Temporal Cluster logs a warning at every 10,000 Events.

> To prevent "runaway" Workflow Executions you can use the Workflow Execution Timeout and/or the Workflow Run Timeout.

Reviewer 3: ```suggestion
To prevent "runaway" Workflow Executions, you can use the Workflow Execution Timeout, the Workflow Run Timeout, or both.
```

A Workflow Execution Timeout can be used to limit the duration of Workflow Execution Chain, and a Workflow Run Timeout can be used to limit the duration an individual Workflow Execution (Run).

> You can use the [Continue-As-New](/docs/concepts/what-is-continue-as-new) feature to create a new Workflow Execution.

Reviewer: Should we mention that closing previous execution and starting the new one is an atomic operation when continue as new is used?

> The Workflow Execution spawned from Continue-As-New has the same Workflow Id, new Run Id, a fresh Event History, and is passed all the appropriate parameters.

Reviewer 3: ```suggestion
The Workflow Execution spawned from Continue-As-New has the same Workflow Id, a new Run Id, and a fresh Event History and is passed all the appropriate parameters.
```

id: what-is-a-workflow-execution
title: What is a Workflow Execution?
sidebar_label: Workflow Execution
description: A Temporal Workflow Execution is a durable, scalable, reliable, and reactive function execution. It is the main unit of execution of a Temporal Application.
tags:
  - explanation
---

A Temporal Workflow Execution is a durable, reliable, and scalable function execution.
It is the main unit of execution of a [Temporal Application](/docs/concepts/what-is-a-temporal-application).

- [How to spawn a Workflow Execution in Go](/docs/go/how-to-spawn-a-workflow-execution-in-go)
- [How to spawn a Workflow Execution in Java](/docs/java/how-to-spawn-a-workflow-execution-in-java)
- [How to spawn a Workflow Execution in PHP](/docs/php/workflows/#starting-workflows)
- [How to spawn a Workflow Execution in TypeScript](/docs/typescript/workflows#how-to-start-and-cancel-workflows)

> Each Temporal Workflow Execution has exclusive access to its local state, executes concurrently to all other Workflow Executions, and communicates with other Workflow Executions via [Signals](/docs/concepts/what-is-a-signal) and the environment via [Activities](/docs/concepts/what-is-an-activity). While a single Workflow Execution has limits on size and throughput, a Temporal Application can consist of millions to billions of Workflow Executions.

Reviewer 2: ```suggestion
Each Temporal Workflow Execution has exclusive access to its local state. It executes concurrently to all other Workflow Executions, and communicates with other Workflow Executions through [Signals](/docs/concepts/what-is-a-signal) and the environment through [Activities](/docs/concepts/what-is-an-activity). 
While a single Workflow Execution has limits on size and throughput, a Temporal Application can consist of millions to billions of Workflow Executions.
```

**Durability**

Durability is the absence of an imposed time limit.

> A Workflow Execution is durable because it executes a Temporal Workflow Definition, also called a Temporal Workflow Function, your application code, effectively once and to completion—whether your code executes for seconds or years.

Reviewer 2: ```suggestion
A Workflow Execution is durable because it executes a Temporal Workflow Definition (also called a Temporal Workflow Function), your application code, effectively once and to completion—whether your code executes for seconds or years.
```

Reviewer 2: You could also break it down into two sentences so the idea of durability is not lost in additional info. For example: "A Workflow Execution is durable because it executes a Temporal Workflow Definition effectively once and to completion—whether your code executes for seconds or years. The Temporal Workflow Definition is your application code , and is also called a Temporal Workflow Function." 

Or remove the second sentence "The Temporal Workflow Definition is your..." and place it within the "What is a Workflow Definition" and link the Workflow Definition here.

**Reliability**

Reliability is responsiveness in the presence of failure.
A Workflow Execution is reliable, because it is fully recoverable after a failure.
The Temporal Platform ensures the state of the Workflow Execution persists in the face of failures and outages and resumes execution from the latest state.

**Scalability**

Scalability is responsiveness in the presence of load.
A single Workflow Execution is limited in size and throughput but is scalable because it can [Continue-As-New](/docs/concepts/what-is-continue-as-new) in response to load.
A Temporal Application is scalable because the Temporal Platform is capable of supporting millions to billions of Workflow Executions executing concurrently, which is realized by the design and nature of the [Temporal Cluster](/docs/concepts/what-is-a-temporal-cluster) and [Worker Processes](/docs/concepts/what-is-a-worker-process).

### Commands & awaitables

A Workflow Execution does two things:

1. Issue [Commands](/docs/concepts/what-is-a-command)
2. Wait on an Awaitables (often called Futures)

![Command generation and waiting](/diagrams/workflow-execution-progession-simple.svg)

> Commands are issued and Awaitables provided by the use of Workflow APIs in the [Workflow Definition](/docs/concepts/what-is-a-workflow-definition).

Author: ```suggestion
Commands are issued and Awaitables are provided by the use of Workflow APIs in the [Workflow Definition](/docs/concepts/what-is-a-workflow-definition).
```

Commands are generated whenever the Workflow Function is executed.
The Worker Process supervises the Command generation and makes sure that it maps to the current Event History.
(For more information, see [Deterministic constraints](/docs/concepts/what-is-a-workflow-definition/#deterministic-constraints).)
The Worker Process batches the Commands and then suspends progress to send the Commands to the Cluster whenever the Workflow Function reaches a place where it can no longer progress without a result from a Awaitable.

A Workflow Execution may only ever block progress on an Awaitable that is provided through a Temporal SDK API.
Awaitables are provided when using APIs for the following:

- Spawning a [Child Workflow Execution](/docs/concepts/what-is-a-child-workflow-execution): Progress can block on confirmation that the Child Workflow Execution started, and on the result of the Child Workflow Execution.
- Sending a [Signal](/docs/concepts/what-is-a-signal): Progress can block on confirmation that the Signal sent.
- Spawning an [Activity Execution](/docs/concepts/what-is-an-activity-execution): Progress can block on the result of the Activity Execution.

> - Requesting cancellation of another Workflow Execution: Progress can block on confirmation that the other Workflow Execution is cancelled.

Author: @tsurdilo I made a guess on this one. Can you confirm?
Or can it block on confirmation that the request to cancel is sent - or both?

- Awaiting

A Workflow Execution may only ever block progress on an Awaitable that is provided through a Temporal SDK API.

### Status

A Workflow Execution can be either Open or Closed.

![Workflow Execution statuses](/diagrams/workflow-execution-statuses.svg)

**Open**

> - Running: The only Open status for a Workflow Execution, when Running the Workflow Execution is when a Workflow Execution is either actively progressing or suspended because it is waiting on something.

Author: ```suggestion
- Running: The only Open status for a Workflow Execution.
When Running the Workflow Execution is either actively progressing or suspended because it is waiting on something.
```

>   When the Workflow Execution is Running, it is either actively progressing or suspended because it is waiting on something.

Reviewer 2: ```suggestion
  When the Workflow Execution is Running, it is either actively progressing or is waiting on something.
```

![Workflow Execution](/diagrams/workflow-execution-swim-lane-01.svg)

### Time constraints

**Is there a limit to how long Workflows can run?**

No, there is no time constraint on how long a Workflow Execution can be Running.

However, Workflow Executions intended to run indefinitely should be written with some care.
The Temporal Cluster stores the complete Event History for the entire lifecycle of a Workflow Execution.
There is a hard limit of 50,000 Events in a Workflow Execution Event History, as well as a hard limit of 50 MB in terms of size.
The Temporal Cluster logs a warning at every 10,000 Events.
When the Event History reaches 50,000 Events or the size limit of 50MB, the Workflow Execution is forcefully terminated.

To prevent "runaway" Workflow Executions, you can use the Workflow Execution Timeout, the Workflow Run Timeout, or both.
A Workflow Execution Timeout can be used to limit the duration of Workflow Execution Chain, and a Workflow Run Timeout can be used to limit the duration an individual Workflow Execution (Run).

> You can use the [Continue-As-New](/docs/concepts/what-is-continue-as-new) feature to close the current Workflow Execution and create a new Workflow Execution in a single an atomic operation.

Reviewer 2: ```suggestion
You can use the [Continue-As-New](/docs/concepts/what-is-continue-as-new) feature to close the current Workflow Execution and create a new Workflow Execution in a single atomic operation.
```
